libs/gltfio/include/gltfio/FilamentAsset.h

```diff
@@ -0,0 +1,148 @@
+ *
+ * Clients must use ResourceLoader to create Texture objects, compute tangent quaternions, and
+ * upload data into vertex buffers and index buffers.
+ *
+ * TODO: This supports skinning but not morphing.
+ * TODO: Only the default glTF scene is loaded, other glTF scenes are ignored.
+ * TODO: Cameras, extras, and extensions are ignored.
+ */
+class FilamentAsset {
+public:
+
+    /**
+     * Gets the list of entities, one for each glTF node. All of these have a Transform component.
+     * Some of the returned entities may also have a Renderable component.
+     */
+    size_t getEntityCount() const noexcept;
+    const utils::Entity* getEntities() const noexcept;
+
+    /** Gets the transform root for the asset, which has no matching glTF node. */
+    utils::Entity getRoot() const noexcept;
+
+    /** Gets all material instances. These are already bound to renderables. */
+    size_t getMaterialInstanceCount() const noexcept;
+    const filament::MaterialInstance* const* getMaterialInstances() const noexcept;
+
+    /** Gets loading instructions for vertex buffers and index buffers. */
+    size_t getBufferBindingCount() const noexcept;
+    const BufferBinding* getBufferBindings() const noexcept;
+
+    /** Gets loading instructions for textures. */
+    size_t getTextureBindingCount() const noexcept;
+    const TextureBinding* getTextureBindings() const noexcept;
+
+    /** Gets the bounding box computed from the supplied min / max values in glTF accessors. */
+    filament::Aabb getBoundingBox() const noexcept;
+
+    /**
+     * Lazily creates the animation engine or returns it from the cache.
+     * The animator is owned by the asset and should not be manually deleted.
+    */
+    Animator* getAnimator() noexcept;
+
+    /**
+     * Reclaims CPU-side memory for URI strings, binding lists, and raw animation data.
+     *
+     * This should only be called after ResourceLoader::loadResources().
+     * If using Animator, this should be called after getAnimator().
+     */
+    void releaseSourceData() noexcept;
+
+protected:
+    FilamentAsset() noexcept = default;
+    ~FilamentAsset() = default;
+
+public:
+    FilamentAsset(FilamentAsset const&) = delete;
+    FilamentAsset(FilamentAsset&&) = delete;
+    FilamentAsset& operator=(FilamentAsset const&) = delete;
+    FilamentAsset& operator=(FilamentAsset&&) = delete;
+};
+
+/**
+ * BufferBinding is a read-only structure that tells clients how to load a source blob into a
+ * VertexBuffer slot or IndexBuffer.
+ *
+ * Each binding instance corresponds to one of the following:
+ *
+ *  (a) One call to VertexBuffer::setBufferAt().
+ *  (b) One call to IndexBuffer::setBuffer().
+ *
+ */
+struct BufferBinding {
+    const char* uri;    // unique identifier for the source blob
+    uint32_t totalSize; // size in bytes of the source blob at the given URI
+    int bufferIndex;    // only used when the destination is a VertexBuffer
```

Comment:
Could you use `uint8_t` or `uint16_t` here? Will the `bufferIndex` ever be negative?

Reply:
Good catch. Done.